packages/wallets/src/evm/wallet.ts

```diff
@@ -26,48 +68,318 @@ export interface ViemWallet {
 }
 
 export class EVMSmartWallet implements ViemWallet {
+    public readonly publicClient: PublicClient<HttpTransport>;
+
     constructor(
-        public readonly publicClient: PublicClient<HttpTransport>,
         public readonly chain: EVMSmartWalletChain,
         private readonly apiClient: ApiClient,
-        private readonly address: Address
-    ) {}
+        private readonly address: Address,
+        private readonly adminSigner: EVMSigner
+    ) {
+        this.publicClient = createPublicClient({
+            chain: toViemChain(chain),
+            transport: http(),
+        });
+    }
 
-    public async balances() {}
-    public async transactions() {}
-    public async nfts() {}
+    public async balances(tokens: Address[]) {
+        return await this.apiClient.getBalance(this.walletLocator, {
+            chains: [this.chain],
+            tokens,
+        });
+    }
+    public async transactions() {
+        const transactions = await this.apiClient.getTransactions(this.walletLocator);
+        return transactions.transactions.filter((transaction) => transaction.walletType === "evm-smart-wallet");
+    }
+    public async nfts(perPage: number, page: number) {
+        return await this.apiClient.getNfts(this.walletLocator, perPage, page);
+    }
 
     public getAddress() {
         return this.address;
     }
 
-    // biome-ignore lint/suspicious/useAwait: <explanation>
     public async getNonce(parameters?: { key?: bigint | undefined } | undefined): Promise<bigint> {
-        throw new Error("Not implemented");
+        const nonce = await this.publicClient.readContract({
+            abi: entryPointAbi,
+            address: ENTRY_POINT_ADDRESS,
+            functionName: "getNonce",
+            args: [this.address, parameters?.key ?? BigInt(0)],
+        });
+        return nonce;
     }
 
-    // biome-ignore lint/suspicious/useAwait: <explanation>
     public async signMessage(parameters: {
         message: SignableMessage;
     }): Promise<Hex> {
-        throw new Error("Not implemented");
+        if (typeof parameters.message !== "string") {
+            throw new Error("Message must be a string");
+        }
+
+        // Create signature
+        const signatureCreationResponse = await this.apiClient.createSignature(this.walletLocator, {
```

Comment:
We should consider abstracting these very long functions into more declarative utility functions with good naming so the whole flow is easier to follow without the need of comments